server/src/main/java/org/elasticsearch/cluster/coordination/MutableMasterHistory.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * This class represents a node's view of the history of which nodes have been elected master over the last 30 minutes. It is kept in
+ * memory, so when a node comes up it does not have any knowledge of previous master history before that point. This object is updated
+ * if and when the cluster state changes with a new master node.
+ */
+public class MutableMasterHistory implements MasterHistory, ClusterStateListener {
+    volatile List<TimeAndMaster> masterHistory;
+    private final Object mutex;
+    Supplier<Long> nowSupplier; // Can be changed for testing
+
+    public MutableMasterHistory(ThreadPool threadPool, ClusterService clusterService) {
+        this.masterHistory = new ArrayList<>();
+        this.mutex = new Object();
+        this.nowSupplier = threadPool::relativeTimeInMillis;
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        DiscoveryNode currentMaster = event.state().nodes().getMasterNode();
+        DiscoveryNode previousMaster = event.previousState().nodes().getMasterNode();
+        if (currentMaster == null || currentMaster.equals(previousMaster) == false || masterHistory.isEmpty()) {
+            masterHistory.add(new TimeAndMaster(nowSupplier.get(), currentMaster));
```

Comment:
This is still not thread-safe as this is not protected by the mutex (iterating over the collection is also not synchronized).
I also don't think we'd want to synchronise in the cluster change notifier thread. 

What do you think about using a lock free collection like `ConcurrentLinkedQueue` (and avoid using `size` as that not constant time anymore - but we'd only need to work with the head of the queue)

Reply:
> Alternatively, we could move this logic to `Coordinator#become{Candidate,Leader,Follower}`

If we move it there though, we need to also move the `removeOldMasterHistory` logic there right? That seems a little awkward doesn't it? Or I could get rid of `removeOldMasterHistory` altogether, and have each read method strip out history older than 30 minutes in a temporary stack-local copy of the history, and only let `clusterChanged` modify it.

Reply:
Yeah it seems reasonable to me to properly clean up the stale entries only when adding new ones. They're tiny, we can afford to clean them up fairly lazily.

Reply:
Also, don't we have to have some sort of lock for this list between readers and writers (or used a threadsafe collection as Andrei suggests)? Even if I'm just making a local copy of the list (new ArrayList(masterHistory)), there's a chance that something will go wrong with making that copy if I don't have a lock. What if I just use a ReadWriteLock. Since there's only ever one writer thread there won't be any contention for that lock right?

Reply:
Bah yes sorry that's right, I forgot about the readers. This is not going to be performance-critical so anything (simple & correct) should be fine.

Reply:
I'd suggest we go for a lock-free data structure (`ConcurrentLinkedQueue` could be a good example here) if we all agree it meets our requirements.

As for the `ReadWriteLock` (and the `Reentrant...` companion) AFAIK this enables multiple readers to hold the read lock, but if there is one reader holding a read lock and the cluster state applier thread wants to acquire the write lock, it will block until the reader is finished - which I don't think is desirable (there're things to do with fairness and such, but they're loosely obeyed AFAIK)

Reply:
The read lock is only ever held long enough to clone the list of masters (which is basically a System.arrayCopy of a very small array), and it only happens if someone is making a health check. So I don't think we'll run into problems here will we?

Reply:
On reflection I have concerns with using `ConcurrentLinkedQueue` and friends in this sort of situation. They have pretty weak guarantees about which concurrent updates are visible when iterating over them. If the update thread removes stale entries and adds new entries then a reader might see neither the stale entries nor the new ones. Maybe it's ok in this context but it depends on details, requires thinking, and risks subtle breakage in some future refactoring. I'd rather have something simpler that doesn't expose intermediate states and doesn't require thinking so hard. Can we just make a new collection each time and expose it in a volatile field somewhere?

Reply:
Actually I think I'm mistaken there. We do get a ClusterChangedEvent when the node finds out about the current master. I'm not sure why I thought otherwise. I introduced that code to solve some problem...